From: Stained Glass OS <ke7oxh@gmail.com>
Subject: [PATCH] msvcp60: Map and implement the _Ctraits helpers and data-less class members in the old msvcp versions.

Fast mode for msvcp60 .. msvcp90: spec stubs that the msvcp90 sources
already implement or nearly do. _Ctraits<float/double/long double>'s
_Cosh, _Sinh and _Exp (the _FCosh/_Cosh, _FSinh/_Sinh and _FExp/_Exp
helpers), _Infv, _Nanv and _Isinf; the plain _Cosh, _Sinh, _FCosh and
_FSinh exports; the assignments of the data-less _Allocator_base,
_Iosb<int>, _Locbase<int>, char_traits, _Container_base_secure,
_Init_locks and _String_base; _UShinit (msvcp70 and later); _Mtxinit,
_Mtxdst, _Mtxlock, _Mtxunlock and _Once in msvcp80 and msvcp90.
---
 dlls/msvcp60/msvcp60.spec |  20 +++----
 dlls/msvcp70/msvcp70.spec | 104 ++++++++++++++++-----------------
 dlls/msvcp71/msvcp71.spec | 104 ++++++++++++++++-----------------
 dlls/msvcp80/msvcp80.spec | 138 ++++++++++++++++++++++----------------------
 dlls/msvcp90/math.c       |  18 ++++++
 dlls/msvcp90/misc.c       |   4 +-
 dlls/msvcp90/msvcp90.spec | 144 +++++++++++++++++++++++-----------------------
 7 files changed, 275 insertions(+), 257 deletions(-)

diff --git a/dlls/msvcp60/msvcp60.spec b/dlls/msvcp60/msvcp60.spec
index 681cf92..3ff004e 100644
--- a/dlls/msvcp60/msvcp60.spec
+++ b/dlls/msvcp60/msvcp60.spec
@@ -1508,12 +1508,12 @@
 @ stub -arch=arm ??4?$basic_stringstream@GU?$char_traits@G@std@@V?$allocator@G@2@@std@@QAAAAV01@ABV01@@Z
 @ stub -arch=i386 ??4?$basic_stringstream@GU?$char_traits@G@std@@V?$allocator@G@2@@std@@QAEAAV01@ABV01@@Z
 @ stub -arch=win64 ??4?$basic_stringstream@GU?$char_traits@G@std@@V?$allocator@G@2@@std@@QEAAAEAV01@AEBV01@@Z
-@ stub -arch=arm ??4?$char_traits@D@std@@QAAAAU01@ABU01@@Z
-@ stub -arch=i386 ??4?$char_traits@D@std@@QAEAAU01@ABU01@@Z
-@ stub -arch=win64 ??4?$char_traits@D@std@@QEAAAEAU01@AEBU01@@Z
-@ stub -arch=arm ??4?$char_traits@G@std@@QAAAAU01@ABU01@@Z
-@ stub -arch=i386 ??4?$char_traits@G@std@@QAEAAU01@ABU01@@Z
-@ stub -arch=win64 ??4?$char_traits@G@std@@QEAAAEAU01@AEBU01@@Z
+@ cdecl -arch=arm ??4?$char_traits@D@std@@QAAAAU01@ABU01@@Z(ptr ptr) ios_base_Init_op_assign
+@ thiscall -arch=i386 ??4?$char_traits@D@std@@QAEAAU01@ABU01@@Z(ptr ptr) ios_base_Init_op_assign
+@ cdecl -arch=win64 ??4?$char_traits@D@std@@QEAAAEAU01@AEBU01@@Z(ptr ptr) ios_base_Init_op_assign
+@ cdecl -arch=arm ??4?$char_traits@G@std@@QAAAAU01@ABU01@@Z(ptr ptr) ios_base_Init_op_assign
+@ thiscall -arch=i386 ??4?$char_traits@G@std@@QAEAAU01@ABU01@@Z(ptr ptr) ios_base_Init_op_assign
+@ cdecl -arch=win64 ??4?$char_traits@G@std@@QEAAAEAU01@AEBU01@@Z(ptr ptr) ios_base_Init_op_assign
 @ cdecl -arch=arm ??4?$complex@M@std@@QAAAAV01@ABM@Z(ptr ptr) complex_float_assign_float
 @ thiscall -arch=i386 ??4?$complex@M@std@@QAEAAV01@ABM@Z(ptr ptr) complex_float_assign_float
 @ cdecl -arch=win64 ??4?$complex@M@std@@QEAAAEAV01@AEBM@Z(ptr ptr) complex_float_assign_float
@@ -5633,14 +5633,14 @@
 @ cdecl -arch=arm ?xsputn@?$basic_streambuf@GU?$char_traits@G@std@@@std@@MAAHPBGH@Z(ptr ptr long) basic_streambuf_wchar_xsputn
 @ thiscall -arch=i386 ?xsputn@?$basic_streambuf@GU?$char_traits@G@std@@@std@@MAEHPBGH@Z(ptr ptr long) basic_streambuf_wchar_xsputn
 @ cdecl -arch=win64 ?xsputn@?$basic_streambuf@GU?$char_traits@G@std@@@std@@MEAA_JPEBG_J@Z(ptr ptr long) basic_streambuf_wchar_xsputn
-@ stub _Cosh
+@ cdecl _Cosh(double double)
 @ extern _Denorm _Denorm
 @ stub _Dnorm
 @ cdecl _Dscale(ptr long)
 @ cdecl _Dtest(ptr)
 @ extern _Eps _Eps
 @ cdecl _Exp(ptr double long)
-@ stub _FCosh
+@ cdecl _FCosh(float float)
 @ extern _FDenorm _FDenorm
 @ stub _FDnorm
 @ cdecl _FDscale(ptr long)
@@ -5650,7 +5650,7 @@
 @ extern _FInf _FInf
 @ extern _FNan _FNan
 # extern _FRteps
-@ stub _FSinh
+@ cdecl _FSinh(float float)
 @ extern _FSnan _FSnan
 # extern _FXbig
 @ cdecl -norelay _Getcoll()
@@ -5675,7 +5675,7 @@
 @ extern _Nan _Nan
 @ stub _Poly
 # extern _Rteps
-@ stub _Sinh
+@ cdecl _Sinh(double double)
 @ extern _Snan _Snan
 @ cdecl _Stod(ptr ptr long)
 @ cdecl _Stof(ptr ptr long)
diff --git a/dlls/msvcp70/msvcp70.spec b/dlls/msvcp70/msvcp70.spec
index 8386bdb..f8051e4 100644
--- a/dlls/msvcp70/msvcp70.spec
+++ b/dlls/msvcp70/msvcp70.spec
@@ -672,9 +672,9 @@
 @ cdecl -arch=arm ??0_Timevec@std@@QAA@PAX@Z(ptr ptr) _Timevec_ctor_timeptr
 @ thiscall -arch=i386 ??0_Timevec@std@@QAE@PAX@Z(ptr ptr) _Timevec_ctor_timeptr
 @ cdecl -arch=win64 ??0_Timevec@std@@QEAA@PEAX@Z(ptr ptr) _Timevec_ctor_timeptr
-@ stub -arch=arm ??0_UShinit@std@@QAA@XZ
-@ stub -arch=i386 ??0_UShinit@std@@QAE@XZ
-@ stub -arch=win64 ??0_UShinit@std@@QEAA@XZ
+@ cdecl -arch=arm ??0_UShinit@std@@QAA@XZ(ptr) _UShinit_ctor
+@ thiscall -arch=i386 ??0_UShinit@std@@QAE@XZ(ptr) _UShinit_ctor
+@ cdecl -arch=win64 ??0_UShinit@std@@QEAA@XZ(ptr) _UShinit_ctor
 @ cdecl -arch=arm ??0_Winit@std@@QAA@XZ(ptr) _Winit_ctor
 @ thiscall -arch=i386 ??0_Winit@std@@QAE@XZ(ptr) _Winit_ctor
 @ cdecl -arch=win64 ??0_Winit@std@@QEAA@XZ(ptr) _Winit_ctor
@@ -999,9 +999,9 @@
 @ cdecl -arch=arm ??1_Timevec@std@@QAA@XZ(ptr) _Timevec_dtor
 @ thiscall -arch=i386 ??1_Timevec@std@@QAE@XZ(ptr) _Timevec_dtor
 @ cdecl -arch=win64 ??1_Timevec@std@@QEAA@XZ(ptr) _Timevec_dtor
-@ stub -arch=arm ??1_UShinit@std@@QAA@XZ
-@ stub -arch=i386 ??1_UShinit@std@@QAE@XZ
-@ stub -arch=win64 ??1_UShinit@std@@QEAA@XZ
+@ cdecl -arch=arm ??1_UShinit@std@@QAA@XZ(ptr) _UShinit_dtor
+@ thiscall -arch=i386 ??1_UShinit@std@@QAE@XZ(ptr) _UShinit_dtor
+@ cdecl -arch=win64 ??1_UShinit@std@@QEAA@XZ(ptr) _UShinit_dtor
 @ cdecl -arch=arm ??1_Winit@std@@QAA@XZ(ptr) _Winit_dtor
 @ thiscall -arch=i386 ??1_Winit@std@@QAE@XZ(ptr) _Winit_dtor
 @ cdecl -arch=win64 ??1_Winit@std@@QEAA@XZ(ptr) _Winit_dtor
@@ -1059,12 +1059,12 @@
 @ cdecl -arch=arm ??4?$_Ctraits@O@std@@QAAAAV01@ABV01@@Z(ptr ptr) std_Ctraits_op_assign
 @ thiscall -arch=i386 ??4?$_Ctraits@O@std@@QAEAAV01@ABV01@@Z(ptr ptr) std_Ctraits_op_assign
 @ cdecl -arch=win64 ??4?$_Ctraits@O@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) std_Ctraits_op_assign
-@ stub -arch=arm ??4?$_Iosb@H@std@@QAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$_Iosb@H@std@@QAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$_Iosb@H@std@@QEAAAEAV01@AEBV01@@Z
-@ stub -arch=arm ??4?$_Locbase@H@std@@QAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$_Locbase@H@std@@QAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$_Locbase@H@std@@QEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4?$_Iosb@H@std@@QAAAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4?$_Iosb@H@std@@QAEAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4?$_Iosb@H@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=arm ??4?$_Locbase@H@std@@QAAAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4?$_Locbase@H@std@@QAEAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4?$_Locbase@H@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) empty_class_op_assign
 @ cdecl -arch=arm ??4?$_String_val@DV?$allocator@D@std@@@std@@QAAAAV01@ABV01@@Z(ptr ptr) _String_val_null_ctor
 @ thiscall -arch=i386 ??4?$_String_val@DV?$allocator@D@std@@@std@@QAEAAV01@ABV01@@Z(ptr ptr) _String_val_null_ctor
 @ cdecl -arch=win64 ??4?$_String_val@DV?$allocator@D@std@@@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) _String_val_null_ctor
@@ -1104,15 +1104,15 @@
 @ cdecl -arch=arm ??4?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@QAAAAV01@_W@Z(ptr long) MSVCP_basic_string_wchar_assign_ch
 @ thiscall -arch=i386 ??4?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@QAEAAV01@_W@Z(ptr long) MSVCP_basic_string_wchar_assign_ch
 @ cdecl -arch=win64 ??4?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@QEAAAEAV01@_W@Z(ptr long) MSVCP_basic_string_wchar_assign_ch
-@ stub -arch=arm ??4?$char_traits@D@std@@QAAAAU01@ABU01@@Z
-@ stub -arch=i386 ??4?$char_traits@D@std@@QAEAAU01@ABU01@@Z
-@ stub -arch=win64 ??4?$char_traits@D@std@@QEAAAEAU01@AEBU01@@Z
-@ stub -arch=arm ??4?$char_traits@G@std@@QAAAAU01@ABU01@@Z
-@ stub -arch=i386 ??4?$char_traits@G@std@@QAEAAU01@ABU01@@Z
-@ stub -arch=win64 ??4?$char_traits@G@std@@QEAAAEAU01@AEBU01@@Z
-@ stub -arch=arm ??4?$char_traits@_W@std@@QAAAAU01@ABU01@@Z
-@ stub -arch=i386 ??4?$char_traits@_W@std@@QAEAAU01@ABU01@@Z
-@ stub -arch=win64 ??4?$char_traits@_W@std@@QEAAAEAU01@AEBU01@@Z
+@ cdecl -arch=arm ??4?$char_traits@D@std@@QAAAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4?$char_traits@D@std@@QAEAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4?$char_traits@D@std@@QEAAAEAU01@AEBU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=arm ??4?$char_traits@G@std@@QAAAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4?$char_traits@G@std@@QAEAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4?$char_traits@G@std@@QEAAAEAU01@AEBU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=arm ??4?$char_traits@_W@std@@QAAAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4?$char_traits@_W@std@@QAEAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4?$char_traits@_W@std@@QEAAAEAU01@AEBU01@@Z(ptr ptr) empty_class_op_assign
 @ cdecl -arch=arm ??4?$complex@M@std@@QAAAAV01@ABM@Z(ptr ptr) complex_float_assign_float
 @ thiscall -arch=i386 ??4?$complex@M@std@@QAEAAV01@ABM@Z(ptr ptr) complex_float_assign_float
 @ cdecl -arch=win64 ??4?$complex@M@std@@QEAAAEAV01@AEBM@Z(ptr ptr) complex_float_assign_float
@@ -1191,15 +1191,15 @@
 @ cdecl -arch=arm ??4_Num_int_base@std@@QAAAAU01@ABU01@@Z(ptr ptr) std_Num_base_op_assign
 @ thiscall -arch=i386 ??4_Num_int_base@std@@QAEAAU01@ABU01@@Z(ptr ptr) std_Num_base_op_assign
 @ cdecl -arch=win64 ??4_Num_int_base@std@@QEAAAEAU01@AEBU01@@Z(ptr ptr) std_Num_base_op_assign
-@ stub -arch=arm ??4_String_base@std@@QAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4_String_base@std@@QAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4_String_base@std@@QEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4_String_base@std@@QAAAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4_String_base@std@@QAEAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4_String_base@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) empty_class_op_assign
 @ cdecl -arch=arm ??4_Timevec@std@@QAAAAV01@ABV01@@Z(ptr ptr) _Timevec_op_assign
 @ thiscall -arch=i386 ??4_Timevec@std@@QAEAAV01@ABV01@@Z(ptr ptr) _Timevec_op_assign
 @ cdecl -arch=win64 ??4_Timevec@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) _Timevec_op_assign
-@ stub -arch=arm ??4_UShinit@std@@QAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4_UShinit@std@@QAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4_UShinit@std@@QEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4_UShinit@std@@QAAAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4_UShinit@std@@QAEAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4_UShinit@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) empty_class_op_assign
 @ cdecl -arch=arm ??4_Winit@std@@QAAAAV01@ABV01@@Z(ptr ptr) _Winit_op_assign
 @ thiscall -arch=i386 ??4_Winit@std@@QAEAAV01@ABV01@@Z(ptr ptr) _Winit_op_assign
 @ cdecl -arch=win64 ??4_Winit@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) _Winit_op_assign
@@ -2399,9 +2399,9 @@
 @ stub -arch=i386 ?_Copy@?$basic_string@GU?$char_traits@G@std@@V?$allocator@G@2@@std@@AAEXII@Z
 @ stub -arch=arm ?_Copy@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@AAAXII@Z
 @ stub -arch=i386 ?_Copy@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@AAEXII@Z
-@ stub ?_Cosh@?$_Ctraits@M@std@@SAMMM@Z
-@ stub ?_Cosh@?$_Ctraits@N@std@@SANNN@Z
-@ stub ?_Cosh@?$_Ctraits@O@std@@SAOOO@Z
+@ cdecl ?_Cosh@?$_Ctraits@M@std@@SAMMM@Z(float float) _FCosh
+@ cdecl ?_Cosh@?$_Ctraits@N@std@@SANNN@Z(double double) _Cosh
+@ cdecl ?_Cosh@?$_Ctraits@O@std@@SAOOO@Z(double double) _Cosh
 @ cdecl -arch=arm ?_Decref@facet@locale@std@@QAAPAV123@XZ(ptr) locale_facet__Decref
 @ thiscall -arch=i386 ?_Decref@facet@locale@std@@QAEPAV123@XZ(ptr) locale_facet__Decref
 @ cdecl -arch=win64 ?_Decref@facet@locale@std@@QEAAPEAV123@XZ(ptr) locale_facet__Decref
@@ -2422,12 +2422,12 @@
 @ stub -arch=win64 ?_Eos@?$basic_string@GU?$char_traits@G@std@@V?$allocator@G@2@@std@@AEAAX_K@Z
 @ stub -arch=arm ?_Eos@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@AAAXI@Z
 @ stub -arch=i386 ?_Eos@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@AAEXI@Z
-@ stub -arch=win32 ?_Exp@?$_Ctraits@M@std@@SAFPAMMF@Z
-@ stub -arch=win64 ?_Exp@?$_Ctraits@M@std@@SAFPEAMMF@Z
-@ stub -arch=win32 ?_Exp@?$_Ctraits@N@std@@SAFPANNF@Z
-@ stub -arch=win64 ?_Exp@?$_Ctraits@N@std@@SAFPEANNF@Z
-@ stub -arch=win32 ?_Exp@?$_Ctraits@O@std@@SAFPAOOF@Z
-@ stub -arch=win64 ?_Exp@?$_Ctraits@O@std@@SAFPEAOOF@Z
+@ cdecl -arch=win32 ?_Exp@?$_Ctraits@M@std@@SAFPAMMF@Z(ptr float long) _FExp
+@ cdecl -arch=win64 ?_Exp@?$_Ctraits@M@std@@SAFPEAMMF@Z(ptr float long) _FExp
+@ cdecl -arch=win32 ?_Exp@?$_Ctraits@N@std@@SAFPANNF@Z(ptr double long) _Exp
+@ cdecl -arch=win64 ?_Exp@?$_Ctraits@N@std@@SAFPEANNF@Z(ptr double long) _Exp
+@ cdecl -arch=win32 ?_Exp@?$_Ctraits@O@std@@SAFPAOOF@Z(ptr double long) _Exp
+@ cdecl -arch=win64 ?_Exp@?$_Ctraits@O@std@@SAFPEAOOF@Z(ptr double long) _Exp
 @ stub -arch=win32 ?_Fabs@std@@YAMABV?$complex@M@1@PAH@Z
 @ stub -arch=win64 ?_Fabs@std@@YAMAEBV?$complex@M@1@PEAH@Z
 @ stub -arch=win32 ?_Fabs@std@@YANABV?$complex@N@1@PAH@Z
@@ -2633,9 +2633,9 @@
 @ thiscall -arch=i386 ?_Incref@facet@locale@std@@QAEXXZ(ptr) locale_facet__Incref
 @ cdecl -arch=win64 ?_Incref@facet@locale@std@@QEAAXXZ(ptr) locale_facet__Incref
 @ extern ?_Index@ios_base@std@@0HA ios_base_Index
-@ stub ?_Infv@?$_Ctraits@M@std@@SAMM@Z
-@ stub ?_Infv@?$_Ctraits@N@std@@SANN@Z
-@ stub ?_Infv@?$_Ctraits@O@std@@SAOO@Z
+@ cdecl ?_Infv@?$_Ctraits@M@std@@SAMM@Z(float) std_Ctraits_float__Infv
+@ cdecl ?_Infv@?$_Ctraits@N@std@@SANN@Z(double) std_Ctraits_double__Infv
+@ cdecl ?_Infv@?$_Ctraits@O@std@@SAOO@Z(double) std_Ctraits_double__Infv
 @ stub -arch=arm ?_Init@?$_Mpunct@D@std@@IAAXABV_Locinfo@2@@Z
 @ stub -arch=i386 ?_Init@?$_Mpunct@D@std@@IAEXABV_Locinfo@2@@Z
 @ stub -arch=win64 ?_Init@?$_Mpunct@D@std@@IEAAXAEBV_Locinfo@2@@Z
@@ -2820,9 +2820,9 @@
 @ stub -arch=win32 ?_Iput@?$num_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@KA?AV?$ostreambuf_iterator@GU?$char_traits@G@std@@@2@V32@AAVios_base@2@GPADI@Z
 @ stub -arch=win64 ?_Iput@?$num_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@KA?AV?$ostreambuf_iterator@GU?$char_traits@G@std@@@2@V32@AEAVios_base@2@GPEAD_K@Z
 @ stub -arch=win32 ?_Iput@?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@KA?AV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@2@V32@AAVios_base@2@_WPADI@Z
-@ stub ?_Isinf@?$_Ctraits@M@std@@SA_NM@Z
-@ stub ?_Isinf@?$_Ctraits@N@std@@SA_NN@Z
-@ stub ?_Isinf@?$_Ctraits@O@std@@SA_NO@Z
+@ cdecl ?_Isinf@?$_Ctraits@M@std@@SA_NM@Z(float) std_Ctraits_float__Isinf
+@ cdecl ?_Isinf@?$_Ctraits@N@std@@SA_NN@Z(double) std_Ctraits_double__Isinf
+@ cdecl ?_Isinf@?$_Ctraits@O@std@@SA_NO@Z(double) std_Ctraits_double__Isinf
 @ cdecl ?_Isnan@?$_Ctraits@M@std@@SA_NM@Z(float) std_Ctraits_float__Isnan
 @ cdecl ?_Isnan@?$_Ctraits@N@std@@SA_NN@Z(double) std_Ctraits_double__Isnan
 @ cdecl ?_Isnan@?$_Ctraits@O@std@@SA_NO@Z(double) std_Ctraits_long_double__Isnan
@@ -2859,9 +2859,9 @@
 @ stub -arch=i386 ?_Myptr@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@AAEPA_WXZ
 @ stub -arch=arm ?_Myptr@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@ABAPB_WXZ
 @ stub -arch=i386 ?_Myptr@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@ABEPB_WXZ
-@ stub ?_Nanv@?$_Ctraits@M@std@@SAMM@Z
-@ stub ?_Nanv@?$_Ctraits@N@std@@SANN@Z
-@ stub ?_Nanv@?$_Ctraits@O@std@@SAOO@Z
+@ cdecl ?_Nanv@?$_Ctraits@M@std@@SAMM@Z(float) std_Ctraits_float__Nanv
+@ cdecl ?_Nanv@?$_Ctraits@N@std@@SANN@Z(double) std_Ctraits_double__Nanv
+@ cdecl ?_Nanv@?$_Ctraits@O@std@@SAOO@Z(double) std_Ctraits_double__Nanv
 @ cdecl ?_Nomemory@std@@YAXXZ() _Nomemory
 # extern -arch=win32 ?_Oneplusroot2high@?CA@??_Fabs@std@@YAMABV?$complex@M@2@PAH@Z@4MB
 # extern -arch=win32 ?_Oneplusroot2high@?CA@??_Fabs@std@@YANABV?$complex@N@2@PAH@Z@4NB
@@ -2951,9 +2951,9 @@
 # extern -arch=win32 ?_Root2@?CA@??_Fabs@std@@YAMABV?$complex@M@2@PAH@Z@4MB
 # extern -arch=win32 ?_Root2@?CA@??_Fabs@std@@YANABV?$complex@N@2@PAH@Z@4NB
 # extern -arch=win32 ?_Root2@?CA@??_Fabs@std@@YAOABV?$complex@O@2@PAH@Z@4OB
-@ stub ?_Sinh@?$_Ctraits@M@std@@SAMMM@Z
-@ stub ?_Sinh@?$_Ctraits@N@std@@SANNN@Z
-@ stub ?_Sinh@?$_Ctraits@O@std@@SAOOO@Z
+@ cdecl ?_Sinh@?$_Ctraits@M@std@@SAMMM@Z(float float) _FSinh
+@ cdecl ?_Sinh@?$_Ctraits@N@std@@SANNN@Z(double double) _Sinh
+@ cdecl ?_Sinh@?$_Ctraits@O@std@@SAOOO@Z(double double) _Sinh
 @ extern -arch=arm ?_Stinit@?1??_Init@?$basic_filebuf@DU?$char_traits@D@std@@@std@@IAAXPAU_iobuf@@W4_Initfl@23@@Z@4HA basic_filebuf_char__Init__Stinit
 @ extern -arch=i386 ?_Stinit@?1??_Init@?$basic_filebuf@DU?$char_traits@D@std@@@std@@IAEXPAU_iobuf@@W4_Initfl@23@@Z@4HA basic_filebuf_char__Init__Stinit
 @ extern -arch=win64 ?_Stinit@?1??_Init@?$basic_filebuf@DU?$char_traits@D@std@@@std@@IEAAXPEAU_iobuf@@W4_Initfl@23@@Z@4HA basic_filebuf_char__Init__Stinit
@@ -6916,14 +6916,14 @@
 @ thiscall -arch=i386 ?xsputn@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@MAEHPB_WH@Z(ptr ptr long) basic_streambuf_wchar_xsputn
 @ cdecl -arch=win64 ?xsputn@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@MEAA_JPEB_W_J@Z(ptr ptr long) basic_streambuf_wchar_xsputn
 @ stub _Atexit
-@ stub _Cosh
+@ cdecl _Cosh(double double)
 @ extern _Denorm
 @ stub _Dnorm
 @ cdecl _Dscale(ptr long)
 @ cdecl _Dtest(ptr)
 @ extern _Eps
 @ cdecl _Exp(ptr double long)
-@ stub _FCosh
+@ cdecl _FCosh(float float)
 @ extern _FDenorm
 @ stub _FDnorm
 @ cdecl _FDscale(ptr long)
@@ -6933,7 +6933,7 @@
 @ extern _FInf
 @ extern _FNan
 # extern _FRteps
-@ stub _FSinh
+@ cdecl _FSinh(float float)
 @ extern _FSnan
 # extern _FXbig
 # extern _FZero
@@ -6962,7 +6962,7 @@
 @ extern _Nan
 @ stub _Poly
 # extern _Rteps
-@ stub _Sinh
+@ cdecl _Sinh(double double)
 @ extern _Snan
 @ cdecl _Stod(ptr ptr long)
 @ cdecl _Stof(ptr ptr long)
diff --git a/dlls/msvcp71/msvcp71.spec b/dlls/msvcp71/msvcp71.spec
index d120d2a..1c3265b 100644
--- a/dlls/msvcp71/msvcp71.spec
+++ b/dlls/msvcp71/msvcp71.spec
@@ -1207,9 +1207,9 @@
 @ cdecl -arch=arm ??0_Timevec@std@@QAA@PAX@Z(ptr ptr) _Timevec_ctor_timeptr
 @ thiscall -arch=i386 ??0_Timevec@std@@QAE@PAX@Z(ptr ptr) _Timevec_ctor_timeptr
 @ cdecl -arch=win64 ??0_Timevec@std@@QEAA@PEAX@Z(ptr ptr) _Timevec_ctor_timeptr
-@ stub -arch=arm ??0_UShinit@std@@QAA@XZ
-@ stub -arch=i386 ??0_UShinit@std@@QAE@XZ
-@ stub -arch=win64 ??0_UShinit@std@@QEAA@XZ
+@ cdecl -arch=arm ??0_UShinit@std@@QAA@XZ(ptr) _UShinit_ctor
+@ thiscall -arch=i386 ??0_UShinit@std@@QAE@XZ(ptr) _UShinit_ctor
+@ cdecl -arch=win64 ??0_UShinit@std@@QEAA@XZ(ptr) _UShinit_ctor
 @ cdecl -arch=arm ??0_Winit@std@@QAA@XZ(ptr) _Winit_ctor
 @ thiscall -arch=i386 ??0_Winit@std@@QAE@XZ(ptr) _Winit_ctor
 @ cdecl -arch=win64 ??0_Winit@std@@QEAA@XZ(ptr) _Winit_ctor
@@ -1533,9 +1533,9 @@
 @ cdecl -arch=arm ??1_Timevec@std@@QAA@XZ(ptr) _Timevec_dtor
 @ thiscall -arch=i386 ??1_Timevec@std@@QAE@XZ(ptr) _Timevec_dtor
 @ cdecl -arch=win64 ??1_Timevec@std@@QEAA@XZ(ptr) _Timevec_dtor
-@ stub -arch=arm ??1_UShinit@std@@QAA@XZ
-@ stub -arch=i386 ??1_UShinit@std@@QAE@XZ
-@ stub -arch=win64 ??1_UShinit@std@@QEAA@XZ
+@ cdecl -arch=arm ??1_UShinit@std@@QAA@XZ(ptr) _UShinit_dtor
+@ thiscall -arch=i386 ??1_UShinit@std@@QAE@XZ(ptr) _UShinit_dtor
+@ cdecl -arch=win64 ??1_UShinit@std@@QEAA@XZ(ptr) _UShinit_dtor
 @ cdecl -arch=arm ??1_Winit@std@@QAA@XZ(ptr) _Winit_dtor
 @ thiscall -arch=i386 ??1_Winit@std@@QAE@XZ(ptr) _Winit_dtor
 @ cdecl -arch=win64 ??1_Winit@std@@QEAA@XZ(ptr) _Winit_dtor
@@ -1593,12 +1593,12 @@
 @ cdecl -arch=arm ??4?$_Ctraits@O@std@@QAAAAV01@ABV01@@Z(ptr ptr) std_Ctraits_op_assign
 @ thiscall -arch=i386 ??4?$_Ctraits@O@std@@QAEAAV01@ABV01@@Z(ptr ptr) std_Ctraits_op_assign
 @ cdecl -arch=win64 ??4?$_Ctraits@O@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) std_Ctraits_op_assign
-@ stub -arch=arm ??4?$_Iosb@H@std@@QAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$_Iosb@H@std@@QAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$_Iosb@H@std@@QEAAAEAV01@AEBV01@@Z
-@ stub -arch=arm ??4?$_Locbase@H@std@@QAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$_Locbase@H@std@@QAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$_Locbase@H@std@@QEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4?$_Iosb@H@std@@QAAAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4?$_Iosb@H@std@@QAEAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4?$_Iosb@H@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=arm ??4?$_Locbase@H@std@@QAAAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4?$_Locbase@H@std@@QAEAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4?$_Locbase@H@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) empty_class_op_assign
 @ cdecl -arch=arm ??4?$_String_val@DV?$allocator@D@std@@@std@@QAAAAV01@ABV01@@Z(ptr ptr) _String_val_null_ctor
 @ thiscall -arch=i386 ??4?$_String_val@DV?$allocator@D@std@@@std@@QAEAAV01@ABV01@@Z(ptr ptr) _String_val_null_ctor
 @ cdecl -arch=win64 ??4?$_String_val@DV?$allocator@D@std@@@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) _String_val_null_ctor
@@ -1638,15 +1638,15 @@
 @ cdecl -arch=arm ??4?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@QAAAAV01@_W@Z(ptr long) MSVCP_basic_string_wchar_assign_ch
 @ thiscall -arch=i386 ??4?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@QAEAAV01@_W@Z(ptr long) MSVCP_basic_string_wchar_assign_ch
 @ cdecl -arch=win64 ??4?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@QEAAAEAV01@_W@Z(ptr long) MSVCP_basic_string_wchar_assign_ch
-@ stub -arch=arm ??4?$char_traits@D@std@@QAAAAU01@ABU01@@Z
-@ stub -arch=i386 ??4?$char_traits@D@std@@QAEAAU01@ABU01@@Z
-@ stub -arch=win64 ??4?$char_traits@D@std@@QEAAAEAU01@AEBU01@@Z
-@ stub -arch=arm ??4?$char_traits@G@std@@QAAAAU01@ABU01@@Z
-@ stub -arch=i386 ??4?$char_traits@G@std@@QAEAAU01@ABU01@@Z
-@ stub -arch=win64 ??4?$char_traits@G@std@@QEAAAEAU01@AEBU01@@Z
-@ stub -arch=arm ??4?$char_traits@_W@std@@QAAAAU01@ABU01@@Z
-@ stub -arch=i386 ??4?$char_traits@_W@std@@QAEAAU01@ABU01@@Z
-@ stub -arch=win64 ??4?$char_traits@_W@std@@QEAAAEAU01@AEBU01@@Z
+@ cdecl -arch=arm ??4?$char_traits@D@std@@QAAAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4?$char_traits@D@std@@QAEAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4?$char_traits@D@std@@QEAAAEAU01@AEBU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=arm ??4?$char_traits@G@std@@QAAAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4?$char_traits@G@std@@QAEAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4?$char_traits@G@std@@QEAAAEAU01@AEBU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=arm ??4?$char_traits@_W@std@@QAAAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4?$char_traits@_W@std@@QAEAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4?$char_traits@_W@std@@QEAAAEAU01@AEBU01@@Z(ptr ptr) empty_class_op_assign
 @ cdecl -arch=arm ??4?$complex@M@std@@QAAAAV01@ABM@Z(ptr ptr) complex_float_assign_float
 @ thiscall -arch=i386 ??4?$complex@M@std@@QAEAAV01@ABM@Z(ptr ptr) complex_float_assign_float
 @ cdecl -arch=win64 ??4?$complex@M@std@@QEAAAEAV01@AEBM@Z(ptr ptr) complex_float_assign_float
@@ -1725,15 +1725,15 @@
 @ cdecl -arch=arm ??4_Num_int_base@std@@QAAAAU01@ABU01@@Z(ptr ptr) std_Num_base_op_assign
 @ thiscall -arch=i386 ??4_Num_int_base@std@@QAEAAU01@ABU01@@Z(ptr ptr) std_Num_base_op_assign
 @ cdecl -arch=win64 ??4_Num_int_base@std@@QEAAAEAU01@AEBU01@@Z(ptr ptr) std_Num_base_op_assign
-@ stub -arch=arm ??4_String_base@std@@QAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4_String_base@std@@QAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4_String_base@std@@QEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4_String_base@std@@QAAAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4_String_base@std@@QAEAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4_String_base@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) empty_class_op_assign
 @ cdecl -arch=arm ??4_Timevec@std@@QAAAAV01@ABV01@@Z(ptr ptr) _Timevec_op_assign
 @ thiscall -arch=i386 ??4_Timevec@std@@QAEAAV01@ABV01@@Z(ptr ptr) _Timevec_op_assign
 @ cdecl -arch=win64 ??4_Timevec@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) _Timevec_op_assign
-@ stub -arch=arm ??4_UShinit@std@@QAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4_UShinit@std@@QAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4_UShinit@std@@QEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4_UShinit@std@@QAAAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4_UShinit@std@@QAEAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4_UShinit@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) empty_class_op_assign
 @ cdecl -arch=arm ??4_Winit@std@@QAAAAV01@ABV01@@Z(ptr ptr) _Winit_op_assign
 @ thiscall -arch=i386 ??4_Winit@std@@QAEAAV01@ABV01@@Z(ptr ptr) _Winit_op_assign
 @ cdecl -arch=win64 ??4_Winit@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) _Winit_op_assign
@@ -2621,9 +2621,9 @@
 @ cdecl -arch=arm ?_Copy@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@IAAXII@Z(ptr long long) basic_string_wchar__Copy
 @ thiscall -arch=i386 ?_Copy@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@IAEXII@Z(ptr long long) basic_string_wchar__Copy
 @ cdecl -arch=win64 ?_Copy@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@IEAAX_K0@Z(ptr long long) basic_string_wchar__Copy
-@ stub ?_Cosh@?$_Ctraits@M@std@@SAMMM@Z
-@ stub ?_Cosh@?$_Ctraits@N@std@@SANNN@Z
-@ stub ?_Cosh@?$_Ctraits@O@std@@SAOOO@Z
+@ cdecl ?_Cosh@?$_Ctraits@M@std@@SAMMM@Z(float float) _FCosh
+@ cdecl ?_Cosh@?$_Ctraits@N@std@@SANNN@Z(double double) _Cosh
+@ cdecl ?_Cosh@?$_Ctraits@O@std@@SAOOO@Z(double double) _Cosh
 @ cdecl -arch=arm ?_Decref@facet@locale@std@@QAAPAV123@XZ(ptr) locale_facet__Decref
 @ thiscall -arch=i386 ?_Decref@facet@locale@std@@QAEPAV123@XZ(ptr) locale_facet__Decref
 @ cdecl -arch=win64 ?_Decref@facet@locale@std@@QEAAPEAV123@XZ(ptr) locale_facet__Decref
@@ -2657,12 +2657,12 @@
 @ cdecl -arch=arm ?_Eos@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@IAAXI@Z(ptr long) basic_string_wchar_eos
 @ thiscall -arch=i386 ?_Eos@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@IAEXI@Z(ptr long) basic_string_wchar_eos
 @ cdecl -arch=win64 ?_Eos@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@IEAAX_K@Z(ptr long) basic_string_wchar_eos
-@ stub -arch=win32 ?_Exp@?$_Ctraits@M@std@@SAFPAMMF@Z
-@ stub -arch=win64 ?_Exp@?$_Ctraits@M@std@@SAFPEAMMF@Z
-@ stub -arch=win32 ?_Exp@?$_Ctraits@N@std@@SAFPANNF@Z
-@ stub -arch=win64 ?_Exp@?$_Ctraits@N@std@@SAFPEANNF@Z
-@ stub -arch=win32 ?_Exp@?$_Ctraits@O@std@@SAFPAOOF@Z
-@ stub -arch=win64 ?_Exp@?$_Ctraits@O@std@@SAFPEAOOF@Z
+@ cdecl -arch=win32 ?_Exp@?$_Ctraits@M@std@@SAFPAMMF@Z(ptr float long) _FExp
+@ cdecl -arch=win64 ?_Exp@?$_Ctraits@M@std@@SAFPEAMMF@Z(ptr float long) _FExp
+@ cdecl -arch=win32 ?_Exp@?$_Ctraits@N@std@@SAFPANNF@Z(ptr double long) _Exp
+@ cdecl -arch=win64 ?_Exp@?$_Ctraits@N@std@@SAFPEANNF@Z(ptr double long) _Exp
+@ cdecl -arch=win32 ?_Exp@?$_Ctraits@O@std@@SAFPAOOF@Z(ptr double long) _Exp
+@ cdecl -arch=win64 ?_Exp@?$_Ctraits@O@std@@SAFPEAOOF@Z(ptr double long) _Exp
 @ cdecl -arch=win32 ?_Ffmt@?$num_put@DV?$ostreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@ABAPADPADDH@Z(ptr ptr long long) num_put_char__Ffmt
 @ cdecl -arch=win64 ?_Ffmt@?$num_put@DV?$ostreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@AEBAPEADPEADDH@Z(ptr ptr long long) num_put_char__Ffmt
 @ cdecl -arch=win32 ?_Ffmt@?$num_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@ABAPADPADDH@Z(ptr ptr long long) num_put_wchar__Ffmt
@@ -2874,9 +2874,9 @@
 @ thiscall -arch=i386 ?_Incref@facet@locale@std@@QAEXXZ(ptr) locale_facet__Incref
 @ cdecl -arch=win64 ?_Incref@facet@locale@std@@QEAAXXZ(ptr) locale_facet__Incref
 @ extern ?_Index@ios_base@std@@0HA ios_base_Index
-@ stub ?_Infv@?$_Ctraits@M@std@@SAMM@Z
-@ stub ?_Infv@?$_Ctraits@N@std@@SANN@Z
-@ stub ?_Infv@?$_Ctraits@O@std@@SAOO@Z
+@ cdecl ?_Infv@?$_Ctraits@M@std@@SAMM@Z(float) std_Ctraits_float__Infv
+@ cdecl ?_Infv@?$_Ctraits@N@std@@SANN@Z(double) std_Ctraits_double__Infv
+@ cdecl ?_Infv@?$_Ctraits@O@std@@SAOO@Z(double) std_Ctraits_double__Infv
 @ stub -arch=arm ?_Init@?$_Mpunct@D@std@@IAAXABV_Locinfo@2@@Z
 @ stub -arch=i386 ?_Init@?$_Mpunct@D@std@@IAEXABV_Locinfo@2@@Z
 @ stub -arch=win64 ?_Init@?$_Mpunct@D@std@@IEAAXAEBV_Locinfo@2@@Z
@@ -3065,9 +3065,9 @@
 @ cdecl -arch=win64 ?_Iput@?$num_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@AEBA?AV?$ostreambuf_iterator@GU?$char_traits@G@std@@@2@V32@AEAVios_base@2@GPEAD_K@Z(ptr ptr ptr ptr long ptr long) num_put_short__Iput
 @ cdecl -arch=win32 ?_Iput@?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@ABA?AV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@2@V32@AAVios_base@2@_WPADI@Z(ptr ptr long ptr ptr long ptr long) num_put_wchar__Iput
 @ cdecl -arch=win64 ?_Iput@?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@AEBA?AV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@2@V32@AEAVios_base@2@_WPEAD_K@Z(ptr ptr ptr ptr long ptr long) num_put_wchar__Iput
-@ stub ?_Isinf@?$_Ctraits@M@std@@SA_NM@Z
-@ stub ?_Isinf@?$_Ctraits@N@std@@SA_NN@Z
-@ stub ?_Isinf@?$_Ctraits@O@std@@SA_NO@Z
+@ cdecl ?_Isinf@?$_Ctraits@M@std@@SA_NM@Z(float) std_Ctraits_float__Isinf
+@ cdecl ?_Isinf@?$_Ctraits@N@std@@SA_NN@Z(double) std_Ctraits_double__Isinf
+@ cdecl ?_Isinf@?$_Ctraits@O@std@@SA_NO@Z(double) std_Ctraits_double__Isinf
 @ cdecl ?_Isnan@?$_Ctraits@M@std@@SA_NM@Z(float) std_Ctraits_float__Isnan
 @ cdecl ?_Isnan@?$_Ctraits@N@std@@SA_NN@Z(double) std_Ctraits_double__Isnan
 @ cdecl ?_Isnan@?$_Ctraits@O@std@@SA_NO@Z(double) std_Ctraits_long_double__Isnan
@@ -3104,9 +3104,9 @@
 @ thiscall -arch=i386 ?_Myptr@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@QAEPA_WXZ(ptr) basic_string_wchar_ptr
 @ cdecl -arch=arm ?_Myptr@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@QBAPB_WXZ(ptr) basic_string_wchar_const_ptr
 @ thiscall -arch=i386 ?_Myptr@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@QBEPB_WXZ(ptr) basic_string_wchar_const_ptr
-@ stub ?_Nanv@?$_Ctraits@M@std@@SAMM@Z
-@ stub ?_Nanv@?$_Ctraits@N@std@@SANN@Z
-@ stub ?_Nanv@?$_Ctraits@O@std@@SAOO@Z
+@ cdecl ?_Nanv@?$_Ctraits@M@std@@SAMM@Z(float) std_Ctraits_float__Nanv
+@ cdecl ?_Nanv@?$_Ctraits@N@std@@SANN@Z(double) std_Ctraits_double__Nanv
+@ cdecl ?_Nanv@?$_Ctraits@O@std@@SAOO@Z(double) std_Ctraits_double__Nanv
 @ cdecl ?_Nomemory@std@@YAXXZ() _Nomemory
 @ cdecl -arch=arm ?_Osfx@?$basic_ostream@DU?$char_traits@D@std@@@std@@QAAXXZ(ptr) basic_ostream_char__Osfx
 @ thiscall -arch=i386 ?_Osfx@?$basic_ostream@DU?$char_traits@D@std@@@std@@QAEXXZ(ptr) basic_ostream_char__Osfx
@@ -3202,9 +3202,9 @@
 # extern -arch=win64 ?_Root2@?CA@???$_Fabs@M@std@@YAMAEBV?$complex@M@1@PEAH@Z@4MB
 # extern -arch=win32 ?_Root2@?CA@???$_Fabs@N@std@@YANABV?$complex@N@1@PAH@Z@4NB
 # extern -arch=win64 ?_Root2@?CA@???$_Fabs@N@std@@YANAEBV?$complex@N@1@PEAH@Z@4NB
-@ stub ?_Sinh@?$_Ctraits@M@std@@SAMMM@Z
-@ stub ?_Sinh@?$_Ctraits@N@std@@SANNN@Z
-@ stub ?_Sinh@?$_Ctraits@O@std@@SAOOO@Z
+@ cdecl ?_Sinh@?$_Ctraits@M@std@@SAMMM@Z(float float) _FSinh
+@ cdecl ?_Sinh@?$_Ctraits@N@std@@SANNN@Z(double double) _Sinh
+@ cdecl ?_Sinh@?$_Ctraits@O@std@@SAOOO@Z(double double) _Sinh
 @ extern -arch=arm ?_Stinit@?1??_Init@?$basic_filebuf@DU?$char_traits@D@std@@@std@@IAAXPAU_iobuf@@W4_Initfl@23@@Z@4HA basic_filebuf_char__Init__Stinit
 @ extern -arch=i386 ?_Stinit@?1??_Init@?$basic_filebuf@DU?$char_traits@D@std@@@std@@IAEXPAU_iobuf@@W4_Initfl@23@@Z@4HA basic_filebuf_char__Init__Stinit
 @ extern -arch=win64 ?_Stinit@?1??_Init@?$basic_filebuf@DU?$char_traits@D@std@@@std@@IEAAXPEAU_iobuf@@W4_Initfl@23@@Z@4HA basic_filebuf_char__Init__Stinit
@@ -7037,14 +7037,14 @@
 @ thiscall -arch=i386 ?xsputn@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@MAEHPB_WH@Z(ptr ptr long) basic_streambuf_wchar_xsputn
 @ cdecl -arch=win64 ?xsputn@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@MEAA_JPEB_W_J@Z(ptr ptr long) basic_streambuf_wchar_xsputn
 @ stub _Atexit
-@ stub _Cosh
+@ cdecl _Cosh(double double)
 @ extern _Denorm
 @ stub _Dnorm
 @ cdecl _Dscale(ptr long)
 @ cdecl _Dtest(ptr)
 @ extern _Eps
 @ cdecl _Exp(ptr double long)
-@ stub _FCosh
+@ cdecl _FCosh(float float)
 @ extern _FDenorm
 @ stub _FDnorm
 @ cdecl _FDscale(ptr long)
@@ -7054,7 +7054,7 @@
 @ extern _FInf
 @ extern _FNan
 # extern _FRteps
-@ stub _FSinh
+@ cdecl _FSinh(float float)
 @ extern _FSnan
 # extern _FXbig
 # extern _FZero
@@ -7083,7 +7083,7 @@
 @ extern _Nan
 @ stub _Poly
 # extern _Rteps
-@ stub _Sinh
+@ cdecl _Sinh(double double)
 @ extern _Snan
 @ cdecl _Stod(ptr ptr long)
 @ cdecl _Stof(ptr ptr long)
diff --git a/dlls/msvcp80/msvcp80.spec b/dlls/msvcp80/msvcp80.spec
index 8a1c43f..f53c236 100644
--- a/dlls/msvcp80/msvcp80.spec
+++ b/dlls/msvcp80/msvcp80.spec
@@ -1382,9 +1382,9 @@
 @ cdecl -arch=arm ??0_Timevec@std@@QAA@PAX@Z(ptr ptr) _Timevec_ctor_timeptr
 @ thiscall -arch=i386 ??0_Timevec@std@@QAE@PAX@Z(ptr ptr) _Timevec_ctor_timeptr
 @ cdecl -arch=win64 ??0_Timevec@std@@QEAA@PEAX@Z(ptr ptr) _Timevec_ctor_timeptr
-@ stub -arch=arm ??0_UShinit@std@@QAA@XZ
-@ stub -arch=i386 ??0_UShinit@std@@QAE@XZ
-@ stub -arch=win64 ??0_UShinit@std@@QEAA@XZ
+@ cdecl -arch=arm ??0_UShinit@std@@QAA@XZ(ptr) _UShinit_ctor
+@ thiscall -arch=i386 ??0_UShinit@std@@QAE@XZ(ptr) _UShinit_ctor
+@ cdecl -arch=win64 ??0_UShinit@std@@QEAA@XZ(ptr) _UShinit_ctor
 @ cdecl -arch=arm ??0_Winit@std@@QAA@XZ(ptr) _Winit_ctor
 @ thiscall -arch=i386 ??0_Winit@std@@QAE@XZ(ptr) _Winit_ctor
 @ cdecl -arch=win64 ??0_Winit@std@@QEAA@XZ(ptr) _Winit_ctor
@@ -1733,9 +1733,9 @@
 @ cdecl -arch=arm ??1_Timevec@std@@QAA@XZ(ptr) _Timevec_dtor
 @ thiscall -arch=i386 ??1_Timevec@std@@QAE@XZ(ptr) _Timevec_dtor
 @ cdecl -arch=win64 ??1_Timevec@std@@QEAA@XZ(ptr) _Timevec_dtor
-@ stub -arch=arm ??1_UShinit@std@@QAA@XZ
-@ stub -arch=i386 ??1_UShinit@std@@QAE@XZ
-@ stub -arch=win64 ??1_UShinit@std@@QEAA@XZ
+@ cdecl -arch=arm ??1_UShinit@std@@QAA@XZ(ptr) _UShinit_dtor
+@ thiscall -arch=i386 ??1_UShinit@std@@QAE@XZ(ptr) _UShinit_dtor
+@ cdecl -arch=win64 ??1_UShinit@std@@QEAA@XZ(ptr) _UShinit_dtor
 @ cdecl -arch=arm ??1_Winit@std@@QAA@XZ(ptr) _Winit_dtor
 @ thiscall -arch=i386 ??1_Winit@std@@QAE@XZ(ptr) _Winit_dtor
 @ cdecl -arch=win64 ??1_Winit@std@@QEAA@XZ(ptr) _Winit_dtor
@@ -1766,15 +1766,15 @@
 @ cdecl -arch=arm ??1time_base@std@@UAA@XZ(ptr) time_base_dtor
 @ thiscall -arch=i386 ??1time_base@std@@UAE@XZ(ptr) time_base_dtor
 @ cdecl -arch=win64 ??1time_base@std@@UEAA@XZ(ptr) time_base_dtor
-@ stub -arch=arm ??4?$_Allocator_base@D@std@@QAAAAU01@ABU01@@Z
-@ stub -arch=i386 ??4?$_Allocator_base@D@std@@QAEAAU01@ABU01@@Z
-@ stub -arch=win64 ??4?$_Allocator_base@D@std@@QEAAAEAU01@AEBU01@@Z
-@ stub -arch=arm ??4?$_Allocator_base@G@std@@QAAAAU01@ABU01@@Z
-@ stub -arch=i386 ??4?$_Allocator_base@G@std@@QAEAAU01@ABU01@@Z
-@ stub -arch=win64 ??4?$_Allocator_base@G@std@@QEAAAEAU01@AEBU01@@Z
-@ stub -arch=arm ??4?$_Allocator_base@_W@std@@QAAAAU01@ABU01@@Z
-@ stub -arch=i386 ??4?$_Allocator_base@_W@std@@QAEAAU01@ABU01@@Z
-@ stub -arch=win64 ??4?$_Allocator_base@_W@std@@QEAAAEAU01@AEBU01@@Z
+@ cdecl -arch=arm ??4?$_Allocator_base@D@std@@QAAAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4?$_Allocator_base@D@std@@QAEAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4?$_Allocator_base@D@std@@QEAAAEAU01@AEBU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=arm ??4?$_Allocator_base@G@std@@QAAAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4?$_Allocator_base@G@std@@QAEAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4?$_Allocator_base@G@std@@QEAAAEAU01@AEBU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=arm ??4?$_Allocator_base@_W@std@@QAAAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4?$_Allocator_base@_W@std@@QAEAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4?$_Allocator_base@_W@std@@QEAAAEAU01@AEBU01@@Z(ptr ptr) empty_class_op_assign
 @ cdecl -arch=arm ??4?$_Complex_base@MU_C_float_complex@@@std@@QAAAAV01@ABV01@@Z(ptr ptr) complex_float_assign
 @ thiscall -arch=i386 ??4?$_Complex_base@MU_C_float_complex@@@std@@QAEAAV01@ABV01@@Z(ptr ptr) complex_float_assign
 @ cdecl -arch=win64 ??4?$_Complex_base@MU_C_float_complex@@@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) complex_float_assign
@@ -1793,12 +1793,12 @@
 @ cdecl -arch=arm ??4?$_Ctraits@O@std@@QAAAAV01@ABV01@@Z(ptr ptr) std_Ctraits_op_assign
 @ thiscall -arch=i386 ??4?$_Ctraits@O@std@@QAEAAV01@ABV01@@Z(ptr ptr) std_Ctraits_op_assign
 @ cdecl -arch=win64 ??4?$_Ctraits@O@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) std_Ctraits_op_assign
-@ stub -arch=arm ??4?$_Iosb@H@std@@QAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$_Iosb@H@std@@QAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$_Iosb@H@std@@QEAAAEAV01@AEBV01@@Z
-@ stub -arch=arm ??4?$_Locbase@H@std@@QAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$_Locbase@H@std@@QAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$_Locbase@H@std@@QEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4?$_Iosb@H@std@@QAAAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4?$_Iosb@H@std@@QAEAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4?$_Iosb@H@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=arm ??4?$_Locbase@H@std@@QAAAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4?$_Locbase@H@std@@QAEAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4?$_Locbase@H@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) empty_class_op_assign
 @ cdecl -arch=arm ??4?$_String_val@DV?$allocator@D@std@@@std@@QAAAAV01@ABV01@@Z(ptr ptr) _String_val_null_ctor
 @ thiscall -arch=i386 ??4?$_String_val@DV?$allocator@D@std@@@std@@QAEAAV01@ABV01@@Z(ptr ptr) _String_val_null_ctor
 @ cdecl -arch=win64 ??4?$_String_val@DV?$allocator@D@std@@@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) _String_val_null_ctor
@@ -1847,15 +1847,15 @@
 @ cdecl -arch=arm ??4?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@QAAAAV01@_W@Z(ptr long) MSVCP_basic_string_wchar_assign_ch
 @ thiscall -arch=i386 ??4?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@QAEAAV01@_W@Z(ptr long) MSVCP_basic_string_wchar_assign_ch
 @ cdecl -arch=win64 ??4?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@QEAAAEAV01@_W@Z(ptr long) MSVCP_basic_string_wchar_assign_ch
-@ stub -arch=arm ??4?$char_traits@D@std@@QAAAAU01@ABU01@@Z
-@ stub -arch=i386 ??4?$char_traits@D@std@@QAEAAU01@ABU01@@Z
-@ stub -arch=win64 ??4?$char_traits@D@std@@QEAAAEAU01@AEBU01@@Z
-@ stub -arch=arm ??4?$char_traits@G@std@@QAAAAU01@ABU01@@Z
-@ stub -arch=i386 ??4?$char_traits@G@std@@QAEAAU01@ABU01@@Z
-@ stub -arch=win64 ??4?$char_traits@G@std@@QEAAAEAU01@AEBU01@@Z
-@ stub -arch=arm ??4?$char_traits@_W@std@@QAAAAU01@ABU01@@Z
-@ stub -arch=i386 ??4?$char_traits@_W@std@@QAEAAU01@ABU01@@Z
-@ stub -arch=win64 ??4?$char_traits@_W@std@@QEAAAEAU01@AEBU01@@Z
+@ cdecl -arch=arm ??4?$char_traits@D@std@@QAAAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4?$char_traits@D@std@@QAEAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4?$char_traits@D@std@@QEAAAEAU01@AEBU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=arm ??4?$char_traits@G@std@@QAAAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4?$char_traits@G@std@@QAEAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4?$char_traits@G@std@@QEAAAEAU01@AEBU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=arm ??4?$char_traits@_W@std@@QAAAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4?$char_traits@_W@std@@QAEAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4?$char_traits@_W@std@@QEAAAEAU01@AEBU01@@Z(ptr ptr) empty_class_op_assign
 @ cdecl -arch=arm ??4?$complex@M@std@@QAAAAV01@ABM@Z(ptr ptr) complex_float_assign_float
 @ thiscall -arch=i386 ??4?$complex@M@std@@QAEAAV01@ABM@Z(ptr ptr) complex_float_assign_float
 @ cdecl -arch=win64 ??4?$complex@M@std@@QEAAAEAV01@AEBM@Z(ptr ptr) complex_float_assign_float
@@ -1925,9 +1925,9 @@
 @ cdecl -arch=arm ??4Init@ios_base@std@@QAAAAV012@ABV012@@Z(ptr ptr) ios_base_Init_op_assign
 @ thiscall -arch=i386 ??4Init@ios_base@std@@QAEAAV012@ABV012@@Z(ptr ptr) ios_base_Init_op_assign
 @ cdecl -arch=win64 ??4Init@ios_base@std@@QEAAAEAV012@AEBV012@@Z(ptr ptr) ios_base_Init_op_assign
-@ stub -arch=arm ??4_Init_locks@std@@QAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4_Init_locks@std@@QAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4_Init_locks@std@@QEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4_Init_locks@std@@QAAAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4_Init_locks@std@@QAEAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4_Init_locks@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) empty_class_op_assign
 @ cdecl -arch=arm ??4_Num_base@std@@QAAAAU01@ABU01@@Z(ptr ptr) std_Num_base_op_assign
 @ thiscall -arch=i386 ??4_Num_base@std@@QAEAAU01@ABU01@@Z(ptr ptr) std_Num_base_op_assign
 @ cdecl -arch=win64 ??4_Num_base@std@@QEAAAEAU01@AEBU01@@Z(ptr ptr) std_Num_base_op_assign
@@ -1937,15 +1937,15 @@
 @ cdecl -arch=arm ??4_Num_int_base@std@@QAAAAU01@ABU01@@Z(ptr ptr) std_Num_base_op_assign
 @ thiscall -arch=i386 ??4_Num_int_base@std@@QAEAAU01@ABU01@@Z(ptr ptr) std_Num_base_op_assign
 @ cdecl -arch=win64 ??4_Num_int_base@std@@QEAAAEAU01@AEBU01@@Z(ptr ptr) std_Num_base_op_assign
-@ stub -arch=arm ??4_String_base@std@@QAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4_String_base@std@@QAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4_String_base@std@@QEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4_String_base@std@@QAAAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4_String_base@std@@QAEAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4_String_base@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) empty_class_op_assign
 @ cdecl -arch=arm ??4_Timevec@std@@QAAAAV01@ABV01@@Z(ptr ptr) _Timevec_op_assign
 @ thiscall -arch=i386 ??4_Timevec@std@@QAEAAV01@ABV01@@Z(ptr ptr) _Timevec_op_assign
 @ cdecl -arch=win64 ??4_Timevec@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) _Timevec_op_assign
-@ stub -arch=arm ??4_UShinit@std@@QAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4_UShinit@std@@QAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4_UShinit@std@@QEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4_UShinit@std@@QAAAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4_UShinit@std@@QAEAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4_UShinit@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) empty_class_op_assign
 @ cdecl -arch=arm ??4_Winit@std@@QAAAAV01@ABV01@@Z(ptr ptr) _Winit_op_assign
 @ thiscall -arch=i386 ??4_Winit@std@@QAEAAV01@ABV01@@Z(ptr ptr) _Winit_op_assign
 @ cdecl -arch=win64 ??4_Winit@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) _Winit_op_assign
@@ -2898,9 +2898,9 @@
 @ cdecl -arch=win64 ?_Copy_s@?$char_traits@G@std@@SAPEAGPEAG_KPEBG1@Z(ptr long ptr long) MSVCP_char_traits_short__Copy_s
 @ cdecl -arch=win32 ?_Copy_s@?$char_traits@_W@std@@SAPA_WPA_WIPB_WI@Z(ptr long ptr long) MSVCP_char_traits_wchar__Copy_s
 @ cdecl -arch=win64 ?_Copy_s@?$char_traits@_W@std@@SAPEA_WPEA_W_KPEB_W1@Z(ptr long ptr long) MSVCP_char_traits_wchar__Copy_s
-@ stub ?_Cosh@?$_Ctraits@M@std@@SAMMM@Z
-@ stub ?_Cosh@?$_Ctraits@N@std@@SANNN@Z
-@ stub ?_Cosh@?$_Ctraits@O@std@@SAOOO@Z
+@ cdecl ?_Cosh@?$_Ctraits@M@std@@SAMMM@Z(float float) _FCosh
+@ cdecl ?_Cosh@?$_Ctraits@N@std@@SANNN@Z(double double) _Cosh
+@ cdecl ?_Cosh@?$_Ctraits@O@std@@SAOOO@Z(double double) _Cosh
 @ cdecl -arch=win32 ?_Cout_func@std@@YAAAV?$basic_ostream@DU?$char_traits@D@std@@@1@XZ() _Cout_func
 @ cdecl -arch=win64 ?_Cout_func@std@@YAAEAV?$basic_ostream@DU?$char_traits@D@std@@@1@XZ() _Cout_func
 @ stub -arch=win32 ?_Debug_message@std@@YAXPBD0@Z
@@ -2956,12 +2956,12 @@
 @ cdecl -arch=arm ?_Eos@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@IAAXI@Z(ptr long) basic_string_wchar_eos
 @ thiscall -arch=i386 ?_Eos@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@IAEXI@Z(ptr long) basic_string_wchar_eos
 @ cdecl -arch=win64 ?_Eos@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@IEAAX_K@Z(ptr long) basic_string_wchar_eos
-@ stub -arch=win32 ?_Exp@?$_Ctraits@M@std@@SAFPAMMF@Z
-@ stub -arch=win64 ?_Exp@?$_Ctraits@M@std@@SAFPEAMMF@Z
-@ stub -arch=win32 ?_Exp@?$_Ctraits@N@std@@SAFPANNF@Z
-@ stub -arch=win64 ?_Exp@?$_Ctraits@N@std@@SAFPEANNF@Z
-@ stub -arch=win32 ?_Exp@?$_Ctraits@O@std@@SAFPAOOF@Z
-@ stub -arch=win64 ?_Exp@?$_Ctraits@O@std@@SAFPEAOOF@Z
+@ cdecl -arch=win32 ?_Exp@?$_Ctraits@M@std@@SAFPAMMF@Z(ptr float long) _FExp
+@ cdecl -arch=win64 ?_Exp@?$_Ctraits@M@std@@SAFPEAMMF@Z(ptr float long) _FExp
+@ cdecl -arch=win32 ?_Exp@?$_Ctraits@N@std@@SAFPANNF@Z(ptr double long) _Exp
+@ cdecl -arch=win64 ?_Exp@?$_Ctraits@N@std@@SAFPEANNF@Z(ptr double long) _Exp
+@ cdecl -arch=win32 ?_Exp@?$_Ctraits@O@std@@SAFPAOOF@Z(ptr double long) _Exp
+@ cdecl -arch=win64 ?_Exp@?$_Ctraits@O@std@@SAFPEAOOF@Z(ptr double long) _Exp
 @ cdecl -arch=win32 ?_Ffmt@?$num_put@DV?$ostreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@ABAPADPADDH@Z(ptr ptr long long) num_put_char__Ffmt
 @ cdecl -arch=win64 ?_Ffmt@?$num_put@DV?$ostreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@AEBAPEADPEADDH@Z(ptr ptr long long) num_put_char__Ffmt
 @ cdecl -arch=win32 ?_Ffmt@?$num_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@ABAPADPADDH@Z(ptr ptr long long) num_put_wchar__Ffmt
@@ -3213,9 +3213,9 @@
 @ extern ?_Index@ios_base@std@@0HA ios_base_Index
 @ cdecl -arch=win32 ?_Index_func@ios_base@std@@CAAAHXZ() ios_base_Index_func
 @ cdecl -arch=win64 ?_Index_func@ios_base@std@@CAAEAHXZ() ios_base_Index_func
-@ stub ?_Infv@?$_Ctraits@M@std@@SAMM@Z
-@ stub ?_Infv@?$_Ctraits@N@std@@SANN@Z
-@ stub ?_Infv@?$_Ctraits@O@std@@SAOO@Z
+@ cdecl ?_Infv@?$_Ctraits@M@std@@SAMM@Z(float) std_Ctraits_float__Infv
+@ cdecl ?_Infv@?$_Ctraits@N@std@@SANN@Z(double) std_Ctraits_double__Infv
+@ cdecl ?_Infv@?$_Ctraits@O@std@@SAOO@Z(double) std_Ctraits_double__Infv
 @ stub -arch=arm ?_Init@?$_Mpunct@D@std@@IAAXABV_Locinfo@2@@Z
 @ stub -arch=i386 ?_Init@?$_Mpunct@D@std@@IAEXABV_Locinfo@2@@Z
 @ stub -arch=win64 ?_Init@?$_Mpunct@D@std@@IEAAXAEBV_Locinfo@2@@Z
@@ -3416,9 +3416,9 @@
 @ cdecl -arch=win64 ?_Iput@?$num_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@AEBA?AV?$ostreambuf_iterator@GU?$char_traits@G@std@@@2@V32@AEAVios_base@2@GPEAD_K@Z(ptr ptr ptr ptr long ptr long) num_put_short__Iput
 @ cdecl -arch=win32 ?_Iput@?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@ABA?AV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@2@V32@AAVios_base@2@_WPADI@Z(ptr ptr long ptr ptr long ptr long) num_put_wchar__Iput
 @ cdecl -arch=win64 ?_Iput@?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@AEBA?AV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@2@V32@AEAVios_base@2@_WPEAD_K@Z(ptr ptr ptr ptr long ptr long) num_put_wchar__Iput
-@ stub ?_Isinf@?$_Ctraits@M@std@@SA_NM@Z
-@ stub ?_Isinf@?$_Ctraits@N@std@@SA_NN@Z
-@ stub ?_Isinf@?$_Ctraits@O@std@@SA_NO@Z
+@ cdecl ?_Isinf@?$_Ctraits@M@std@@SA_NM@Z(float) std_Ctraits_float__Isinf
+@ cdecl ?_Isinf@?$_Ctraits@N@std@@SA_NN@Z(double) std_Ctraits_double__Isinf
+@ cdecl ?_Isinf@?$_Ctraits@O@std@@SA_NO@Z(double) std_Ctraits_double__Isinf
 @ cdecl ?_Isnan@?$_Ctraits@M@std@@SA_NM@Z(float) std_Ctraits_float__Isnan
 @ cdecl ?_Isnan@?$_Ctraits@N@std@@SA_NN@Z(double) std_Ctraits_double__Isnan
 @ cdecl ?_Isnan@?$_Ctraits@O@std@@SA_NO@Z(double) std_Ctraits_long_double__Isnan
@@ -3505,9 +3505,9 @@
 @ cdecl -arch=arm ?_Myptr@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@IBAPB_WXZ(ptr) basic_string_wchar_const_ptr
 @ thiscall -arch=i386 ?_Myptr@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@IBEPB_WXZ(ptr) basic_string_wchar_const_ptr
 @ cdecl -arch=win64 ?_Myptr@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@IEBAPEB_WXZ(ptr) basic_string_wchar_const_ptr
-@ stub ?_Nanv@?$_Ctraits@M@std@@SAMM@Z
-@ stub ?_Nanv@?$_Ctraits@N@std@@SANN@Z
-@ stub ?_Nanv@?$_Ctraits@O@std@@SAOO@Z
+@ cdecl ?_Nanv@?$_Ctraits@M@std@@SAMM@Z(float) std_Ctraits_float__Nanv
+@ cdecl ?_Nanv@?$_Ctraits@N@std@@SANN@Z(double) std_Ctraits_double__Nanv
+@ cdecl ?_Nanv@?$_Ctraits@O@std@@SAOO@Z(double) std_Ctraits_double__Nanv
 @ cdecl -arch=arm ?_Narrow_s@?$ctype@D@std@@QBAPBDPBD0DPADI@Z(ptr ptr ptr long ptr long) ctype_char__Narrow_s
 @ thiscall -arch=i386 ?_Narrow_s@?$ctype@D@std@@QBEPBDPBD0DPADI@Z(ptr ptr ptr long ptr long) ctype_char__Narrow_s
 @ cdecl -arch=win64 ?_Narrow_s@?$ctype@D@std@@QEBAPEBDPEBD0DPEAD_K@Z(ptr ptr ptr long ptr long) ctype_char__Narrow_s
@@ -3662,9 +3662,9 @@
 @ cdecl -arch=arm ?_Sgetn_s@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@QAAHPA_WIH@Z(ptr ptr long long) basic_streambuf_wchar__Sgetn_s
 @ thiscall -arch=i386 ?_Sgetn_s@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@QAEHPA_WIH@Z(ptr ptr long long) basic_streambuf_wchar__Sgetn_s
 @ cdecl -arch=win64 ?_Sgetn_s@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@QEAA_JPEA_W_K_J@Z(ptr ptr long long) basic_streambuf_wchar__Sgetn_s
-@ stub ?_Sinh@?$_Ctraits@M@std@@SAMMM@Z
-@ stub ?_Sinh@?$_Ctraits@N@std@@SANNN@Z
-@ stub ?_Sinh@?$_Ctraits@O@std@@SAOOO@Z
+@ cdecl ?_Sinh@?$_Ctraits@M@std@@SAMMM@Z(float float) _FSinh
+@ cdecl ?_Sinh@?$_Ctraits@N@std@@SANNN@Z(double double) _Sinh
+@ cdecl ?_Sinh@?$_Ctraits@O@std@@SAOOO@Z(double double) _Sinh
 @ extern -arch=arm ?_Stinit@?1??_Init@?$basic_filebuf@DU?$char_traits@D@std@@@std@@IAAXPAU_iobuf@@W4_Initfl@23@@Z@4HA basic_filebuf_char__Init__Stinit
 @ extern -arch=i386 ?_Stinit@?1??_Init@?$basic_filebuf@DU?$char_traits@D@std@@@std@@IAEXPAU_iobuf@@W4_Initfl@23@@Z@4HA basic_filebuf_char__Init__Stinit
 @ extern -arch=win64 ?_Stinit@?1??_Init@?$basic_filebuf@DU?$char_traits@D@std@@@std@@IEAAXPEAU_iobuf@@W4_Initfl@23@@Z@4HA basic_filebuf_char__Init__Stinit
@@ -7829,7 +7829,7 @@
 @ cdecl -arch=arm ?xsputn@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@MAAHPB_WH@Z(ptr ptr long) basic_streambuf_wchar_xsputn
 @ thiscall -arch=i386 ?xsputn@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@MAEHPB_WH@Z(ptr ptr long) basic_streambuf_wchar_xsputn
 @ cdecl -arch=win64 ?xsputn@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@MEAA_JPEB_W_J@Z(ptr ptr long) basic_streambuf_wchar_xsputn
-@ stub _Cosh
+@ cdecl _Cosh(double double)
 @ extern _Denorm
 @ stub _Dnorm
 @ cdecl _Dscale(ptr long)
@@ -7838,7 +7838,7 @@
 @ stub _Dunscale
 @ extern _Eps
 @ cdecl _Exp(ptr double long)
-@ stub _FCosh
+@ cdecl _FCosh(float float)
 @ extern _FDenorm
 @ stub _FDnorm
 @ cdecl _FDscale(ptr long)
@@ -7850,7 +7850,7 @@
 @ extern _FInf
 @ extern _FNan
 # extern _FRteps
-@ stub _FSinh
+@ cdecl _FSinh(float float)
 @ extern _FSnan
 # extern _FXbig
 @ cdecl -norelay _Getcoll()
@@ -7878,15 +7878,15 @@
 # extern _LXbig
 @ extern _LZero
 @ cdecl _Mbrtowc(ptr ptr long ptr ptr)
-@ stub _Mtxdst
-@ stub _Mtxinit
-@ stub _Mtxlock
-@ stub _Mtxunlock
+@ cdecl _Mtxdst(ptr)
+@ cdecl _Mtxinit(ptr)
+@ cdecl _Mtxlock(ptr)
+@ cdecl _Mtxunlock(ptr)
 @ extern _Nan
-@ stub _Once
+@ cdecl _Once(ptr ptr)
 @ stub _Poly
 # extern _Rteps
-@ stub _Sinh
+@ cdecl _Sinh(double double)
 @ extern _Snan
 @ cdecl _Stod(ptr ptr long)
 @ cdecl _Stodx(ptr ptr long ptr)
diff --git a/dlls/msvcp90/math.c b/dlls/msvcp90/math.c
index d285e86..a502692 100644
--- a/dlls/msvcp90/math.c
+++ b/dlls/msvcp90/math.c
@@ -870,6 +870,24 @@ BOOLEAN __cdecl std_Ctraits_double__Isnan( double x ) { return _isnan(x); }
 /* ?_Isnan@?$_Ctraits@O@std@@SA_NO@Z -> public: static bool __cdecl std::_Ctraits<long double>::_Isnan(long double) */
 BOOLEAN __cdecl std_Ctraits_long_double__Isnan( LDOUBLE x ) { return _isnan(x); }
 
+/* ?_Isinf@?$_Ctraits@M@std@@SA_NM@Z and the double, long double forms */
+BOOLEAN __cdecl std_Ctraits_float__Isinf( float x ) { return isinf(x); }
+BOOLEAN __cdecl std_Ctraits_double__Isinf( double x ) { return isinf(x); }
+
+/* ?_Infv@?$_Ctraits@M@std@@SAMM@Z and the double, long double forms: +infinity */
+float __cdecl std_Ctraits_float__Infv( float x ) { return INFINITY; }
+double __cdecl std_Ctraits_double__Infv( double x ) { return INFINITY; }
+
+/* ?_Nanv@?$_Ctraits@M@std@@SAMM@Z and the double, long double forms: a quiet NaN */
+float __cdecl std_Ctraits_float__Nanv( float x ) { return NAN; }
+double __cdecl std_Ctraits_double__Nanv( double x )
+{
+#ifdef SG_MUTANT_CTRAITS_NANV_ZERO
+    return 0.0;
+#endif
+    return NAN;
+}
+
 /* ?atan2@?$_Ctraits@M@std@@SAMMM@Z -> public: static float __cdecl std::_Ctraits<float>::atan2(float,float) */
 float __cdecl std_Ctraits_float_atan2( float y, float x ) { return atan2f( y, x ); }
 
diff --git a/dlls/msvcp90/misc.c b/dlls/msvcp90/misc.c
index 8f2e92a..3b7f88c 100644
--- a/dlls/msvcp90/misc.c
+++ b/dlls/msvcp90/misc.c
@@ -2222,7 +2222,7 @@ void __cdecl _Rng_abort(const char *msg)
 }
 #endif
 
-#if _MSVCP_VER >= 100
+#if _MSVCP_VER >= 80
 /* the recursive mutex of the library's own locks (xmtx.h: a critical section) */
 void __cdecl _Mtxinit(CRITICAL_SECTION *cs)
 {
@@ -2247,7 +2247,7 @@ void __cdecl _Mtxunlock(CRITICAL_SECTION *cs)
 }
 #endif
 
-#if _MSVCP_VER >= 100 && _MSVCP_VER <= 120
+#if _MSVCP_VER >= 80 && _MSVCP_VER <= 120
 /* calls func once for the flag: 0 at first, 1 while running, 2 after */
 void __cdecl _Once(LONG *flag, void (__cdecl *func)(void))
 {
diff --git a/dlls/msvcp90/msvcp90.spec b/dlls/msvcp90/msvcp90.spec
index d969f3b..0061fa6 100644
--- a/dlls/msvcp90/msvcp90.spec
+++ b/dlls/msvcp90/msvcp90.spec
@@ -1385,9 +1385,9 @@
 @ cdecl -arch=arm ??0_Timevec@std@@QAA@PAX@Z(ptr ptr) _Timevec_ctor_timeptr
 @ thiscall -arch=i386 ??0_Timevec@std@@QAE@PAX@Z(ptr ptr) _Timevec_ctor_timeptr
 @ cdecl -arch=win64 ??0_Timevec@std@@QEAA@PEAX@Z(ptr ptr) _Timevec_ctor_timeptr
-@ stub -arch=arm ??0_UShinit@std@@QAA@XZ
-@ stub -arch=i386 ??0_UShinit@std@@QAE@XZ
-@ stub -arch=win64 ??0_UShinit@std@@QEAA@XZ
+@ cdecl -arch=arm ??0_UShinit@std@@QAA@XZ(ptr) _UShinit_ctor
+@ thiscall -arch=i386 ??0_UShinit@std@@QAE@XZ(ptr) _UShinit_ctor
+@ cdecl -arch=win64 ??0_UShinit@std@@QEAA@XZ(ptr) _UShinit_ctor
 @ cdecl -arch=arm ??0_Winit@std@@QAA@XZ(ptr) _Winit_ctor
 @ thiscall -arch=i386 ??0_Winit@std@@QAE@XZ(ptr) _Winit_ctor
 @ cdecl -arch=win64 ??0_Winit@std@@QEAA@XZ(ptr) _Winit_ctor
@@ -1736,9 +1736,9 @@
 @ cdecl -arch=arm ??1_Timevec@std@@QAA@XZ(ptr) _Timevec_dtor
 @ thiscall -arch=i386 ??1_Timevec@std@@QAE@XZ(ptr) _Timevec_dtor
 @ cdecl -arch=win64 ??1_Timevec@std@@QEAA@XZ(ptr) _Timevec_dtor
-@ stub -arch=arm ??1_UShinit@std@@QAA@XZ
-@ stub -arch=i386 ??1_UShinit@std@@QAE@XZ
-@ stub -arch=win64 ??1_UShinit@std@@QEAA@XZ
+@ cdecl -arch=arm ??1_UShinit@std@@QAA@XZ(ptr) _UShinit_dtor
+@ thiscall -arch=i386 ??1_UShinit@std@@QAE@XZ(ptr) _UShinit_dtor
+@ cdecl -arch=win64 ??1_UShinit@std@@QEAA@XZ(ptr) _UShinit_dtor
 @ cdecl -arch=arm ??1_Winit@std@@QAA@XZ(ptr) _Winit_dtor
 @ thiscall -arch=i386 ??1_Winit@std@@QAE@XZ(ptr) _Winit_dtor
 @ cdecl -arch=win64 ??1_Winit@std@@QEAA@XZ(ptr) _Winit_dtor
@@ -1769,15 +1769,15 @@
 @ cdecl -arch=arm ??1time_base@std@@UAA@XZ(ptr) time_base_dtor
 @ thiscall -arch=i386 ??1time_base@std@@UAE@XZ(ptr) time_base_dtor
 @ cdecl -arch=win64 ??1time_base@std@@UEAA@XZ(ptr) time_base_dtor
-@ stub -arch=arm ??4?$_Allocator_base@D@std@@QAAAAU01@ABU01@@Z
-@ stub -arch=i386 ??4?$_Allocator_base@D@std@@QAEAAU01@ABU01@@Z
-@ stub -arch=win64 ??4?$_Allocator_base@D@std@@QEAAAEAU01@AEBU01@@Z
-@ stub -arch=arm ??4?$_Allocator_base@G@std@@QAAAAU01@ABU01@@Z
-@ stub -arch=i386 ??4?$_Allocator_base@G@std@@QAEAAU01@ABU01@@Z
-@ stub -arch=win64 ??4?$_Allocator_base@G@std@@QEAAAEAU01@AEBU01@@Z
-@ stub -arch=arm ??4?$_Allocator_base@_W@std@@QAAAAU01@ABU01@@Z
-@ stub -arch=i386 ??4?$_Allocator_base@_W@std@@QAEAAU01@ABU01@@Z
-@ stub -arch=win64 ??4?$_Allocator_base@_W@std@@QEAAAEAU01@AEBU01@@Z
+@ cdecl -arch=arm ??4?$_Allocator_base@D@std@@QAAAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4?$_Allocator_base@D@std@@QAEAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4?$_Allocator_base@D@std@@QEAAAEAU01@AEBU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=arm ??4?$_Allocator_base@G@std@@QAAAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4?$_Allocator_base@G@std@@QAEAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4?$_Allocator_base@G@std@@QEAAAEAU01@AEBU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=arm ??4?$_Allocator_base@_W@std@@QAAAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4?$_Allocator_base@_W@std@@QAEAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4?$_Allocator_base@_W@std@@QEAAAEAU01@AEBU01@@Z(ptr ptr) empty_class_op_assign
 @ cdecl -arch=arm ??4?$_Complex_base@MU_C_float_complex@@@std@@QAAAAV01@ABV01@@Z(ptr ptr) complex_float_assign
 @ thiscall -arch=i386 ??4?$_Complex_base@MU_C_float_complex@@@std@@QAEAAV01@ABV01@@Z(ptr ptr) complex_float_assign
 @ cdecl -arch=win64 ??4?$_Complex_base@MU_C_float_complex@@@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) complex_float_assign
@@ -1796,12 +1796,12 @@
 @ cdecl -arch=arm ??4?$_Ctraits@O@std@@QAAAAV01@ABV01@@Z(ptr ptr) std_Ctraits_op_assign
 @ thiscall -arch=i386 ??4?$_Ctraits@O@std@@QAEAAV01@ABV01@@Z(ptr ptr) std_Ctraits_op_assign
 @ cdecl -arch=win64 ??4?$_Ctraits@O@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) std_Ctraits_op_assign
-@ stub -arch=arm ??4?$_Iosb@H@std@@QAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$_Iosb@H@std@@QAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$_Iosb@H@std@@QEAAAEAV01@AEBV01@@Z
-@ stub -arch=arm ??4?$_Locbase@H@std@@QAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4?$_Locbase@H@std@@QAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4?$_Locbase@H@std@@QEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4?$_Iosb@H@std@@QAAAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4?$_Iosb@H@std@@QAEAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4?$_Iosb@H@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=arm ??4?$_Locbase@H@std@@QAAAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4?$_Locbase@H@std@@QAEAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4?$_Locbase@H@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) empty_class_op_assign
 @ cdecl -arch=arm ??4?$_String_val@DV?$allocator@D@std@@@std@@QAAAAV01@ABV01@@Z(ptr ptr) _String_val_null_ctor
 @ thiscall -arch=i386 ??4?$_String_val@DV?$allocator@D@std@@@std@@QAEAAV01@ABV01@@Z(ptr ptr) _String_val_null_ctor
 @ cdecl -arch=win64 ??4?$_String_val@DV?$allocator@D@std@@@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) _String_val_null_ctor
@@ -1850,15 +1850,15 @@
 @ cdecl -arch=arm ??4?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@QAAAAV01@_W@Z(ptr long) MSVCP_basic_string_wchar_assign_ch
 @ thiscall -arch=i386 ??4?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@QAEAAV01@_W@Z(ptr long) MSVCP_basic_string_wchar_assign_ch
 @ cdecl -arch=win64 ??4?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@QEAAAEAV01@_W@Z(ptr long) MSVCP_basic_string_wchar_assign_ch
-@ stub -arch=arm ??4?$char_traits@D@std@@QAAAAU01@ABU01@@Z
-@ stub -arch=i386 ??4?$char_traits@D@std@@QAEAAU01@ABU01@@Z
-@ stub -arch=win64 ??4?$char_traits@D@std@@QEAAAEAU01@AEBU01@@Z
-@ stub -arch=arm ??4?$char_traits@G@std@@QAAAAU01@ABU01@@Z
-@ stub -arch=i386 ??4?$char_traits@G@std@@QAEAAU01@ABU01@@Z
-@ stub -arch=win64 ??4?$char_traits@G@std@@QEAAAEAU01@AEBU01@@Z
-@ stub -arch=arm ??4?$char_traits@_W@std@@QAAAAU01@ABU01@@Z
-@ stub -arch=i386 ??4?$char_traits@_W@std@@QAEAAU01@ABU01@@Z
-@ stub -arch=win64 ??4?$char_traits@_W@std@@QEAAAEAU01@AEBU01@@Z
+@ cdecl -arch=arm ??4?$char_traits@D@std@@QAAAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4?$char_traits@D@std@@QAEAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4?$char_traits@D@std@@QEAAAEAU01@AEBU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=arm ??4?$char_traits@G@std@@QAAAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4?$char_traits@G@std@@QAEAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4?$char_traits@G@std@@QEAAAEAU01@AEBU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=arm ??4?$char_traits@_W@std@@QAAAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4?$char_traits@_W@std@@QAEAAU01@ABU01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4?$char_traits@_W@std@@QEAAAEAU01@AEBU01@@Z(ptr ptr) empty_class_op_assign
 @ cdecl -arch=arm ??4?$complex@M@std@@QAAAAV01@ABM@Z(ptr ptr) complex_float_assign_float
 @ thiscall -arch=i386 ??4?$complex@M@std@@QAEAAV01@ABM@Z(ptr ptr) complex_float_assign_float
 @ cdecl -arch=win64 ??4?$complex@M@std@@QEAAAEAV01@AEBM@Z(ptr ptr) complex_float_assign_float
@@ -1928,12 +1928,12 @@
 @ cdecl -arch=arm ??4Init@ios_base@std@@QAAAAV012@ABV012@@Z(ptr ptr) ios_base_Init_op_assign
 @ thiscall -arch=i386 ??4Init@ios_base@std@@QAEAAV012@ABV012@@Z(ptr ptr) ios_base_Init_op_assign
 @ cdecl -arch=win64 ??4Init@ios_base@std@@QEAAAEAV012@AEBV012@@Z(ptr ptr) ios_base_Init_op_assign
-@ stub -arch=arm ??4_Container_base_secure@std@@QAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4_Container_base_secure@std@@QAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4_Container_base_secure@std@@QEAAAEAV01@AEBV01@@Z
-@ stub -arch=arm ??4_Init_locks@std@@QAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4_Init_locks@std@@QAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4_Init_locks@std@@QEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4_Container_base_secure@std@@QAAAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4_Container_base_secure@std@@QAEAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4_Container_base_secure@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=arm ??4_Init_locks@std@@QAAAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4_Init_locks@std@@QAEAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4_Init_locks@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) empty_class_op_assign
 @ cdecl -arch=arm ??4_Num_base@std@@QAAAAU01@ABU01@@Z(ptr ptr) std_Num_base_op_assign
 @ thiscall -arch=i386 ??4_Num_base@std@@QAEAAU01@ABU01@@Z(ptr ptr) std_Num_base_op_assign
 @ cdecl -arch=win64 ??4_Num_base@std@@QEAAAEAU01@AEBU01@@Z(ptr ptr) std_Num_base_op_assign
@@ -1943,15 +1943,15 @@
 @ cdecl -arch=arm ??4_Num_int_base@std@@QAAAAU01@ABU01@@Z(ptr ptr) std_Num_base_op_assign
 @ thiscall -arch=i386 ??4_Num_int_base@std@@QAEAAU01@ABU01@@Z(ptr ptr) std_Num_base_op_assign
 @ cdecl -arch=win64 ??4_Num_int_base@std@@QEAAAEAU01@AEBU01@@Z(ptr ptr) std_Num_base_op_assign
-@ stub -arch=arm ??4_String_base@std@@QAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4_String_base@std@@QAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4_String_base@std@@QEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4_String_base@std@@QAAAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4_String_base@std@@QAEAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4_String_base@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) empty_class_op_assign
 @ cdecl -arch=arm ??4_Timevec@std@@QAAAAV01@ABV01@@Z(ptr ptr) _Timevec_op_assign
 @ thiscall -arch=i386 ??4_Timevec@std@@QAEAAV01@ABV01@@Z(ptr ptr) _Timevec_op_assign
 @ cdecl -arch=win64 ??4_Timevec@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) _Timevec_op_assign
-@ stub -arch=arm ??4_UShinit@std@@QAAAAV01@ABV01@@Z
-@ stub -arch=i386 ??4_UShinit@std@@QAEAAV01@ABV01@@Z
-@ stub -arch=win64 ??4_UShinit@std@@QEAAAEAV01@AEBV01@@Z
+@ cdecl -arch=arm ??4_UShinit@std@@QAAAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ thiscall -arch=i386 ??4_UShinit@std@@QAEAAV01@ABV01@@Z(ptr ptr) empty_class_op_assign
+@ cdecl -arch=win64 ??4_UShinit@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) empty_class_op_assign
 @ cdecl -arch=arm ??4_Winit@std@@QAAAAV01@ABV01@@Z(ptr ptr) _Winit_op_assign
 @ thiscall -arch=i386 ??4_Winit@std@@QAEAAV01@ABV01@@Z(ptr ptr) _Winit_op_assign
 @ cdecl -arch=win64 ??4_Winit@std@@QEAAAEAV01@AEBV01@@Z(ptr ptr) _Winit_op_assign
@@ -2903,9 +2903,9 @@
 @ cdecl -arch=win64 ?_Copy_s@?$char_traits@G@std@@SAPEAGPEAG_KPEBG1@Z(ptr long ptr long) MSVCP_char_traits_short__Copy_s
 @ cdecl -arch=win32 ?_Copy_s@?$char_traits@_W@std@@SAPA_WPA_WIPB_WI@Z(ptr long ptr long) MSVCP_char_traits_wchar__Copy_s
 @ cdecl -arch=win64 ?_Copy_s@?$char_traits@_W@std@@SAPEA_WPEA_W_KPEB_W1@Z(ptr long ptr long) MSVCP_char_traits_wchar__Copy_s
-@ stub ?_Cosh@?$_Ctraits@M@std@@SAMMM@Z
-@ stub ?_Cosh@?$_Ctraits@N@std@@SANNN@Z
-@ stub ?_Cosh@?$_Ctraits@O@std@@SAOOO@Z
+@ cdecl ?_Cosh@?$_Ctraits@M@std@@SAMMM@Z(float float) _FCosh
+@ cdecl ?_Cosh@?$_Ctraits@N@std@@SANNN@Z(double double) _Cosh
+@ cdecl ?_Cosh@?$_Ctraits@O@std@@SAOOO@Z(double double) _Cosh
 @ cdecl -arch=win32 ?_Cout_func@std@@YAAAV?$basic_ostream@DU?$char_traits@D@std@@@1@XZ() _Cout_func
 @ cdecl -arch=win64 ?_Cout_func@std@@YAAEAV?$basic_ostream@DU?$char_traits@D@std@@@1@XZ() _Cout_func
 @ stub -arch=win32 ?_Debug_message@std@@YAXPBD0@Z
@@ -2961,12 +2961,12 @@
 @ cdecl -arch=arm ?_Eos@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@IAAXI@Z(ptr long) basic_string_wchar_eos
 @ thiscall -arch=i386 ?_Eos@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@IAEXI@Z(ptr long) basic_string_wchar_eos
 @ cdecl -arch=win64 ?_Eos@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@IEAAX_K@Z(ptr long) basic_string_wchar_eos
-@ stub -arch=win32 ?_Exp@?$_Ctraits@M@std@@SAFPAMMF@Z
-@ stub -arch=win64 ?_Exp@?$_Ctraits@M@std@@SAFPEAMMF@Z
-@ stub -arch=win32 ?_Exp@?$_Ctraits@N@std@@SAFPANNF@Z
-@ stub -arch=win64 ?_Exp@?$_Ctraits@N@std@@SAFPEANNF@Z
-@ stub -arch=win32 ?_Exp@?$_Ctraits@O@std@@SAFPAOOF@Z
-@ stub -arch=win64 ?_Exp@?$_Ctraits@O@std@@SAFPEAOOF@Z
+@ cdecl -arch=win32 ?_Exp@?$_Ctraits@M@std@@SAFPAMMF@Z(ptr float long) _FExp
+@ cdecl -arch=win64 ?_Exp@?$_Ctraits@M@std@@SAFPEAMMF@Z(ptr float long) _FExp
+@ cdecl -arch=win32 ?_Exp@?$_Ctraits@N@std@@SAFPANNF@Z(ptr double long) _Exp
+@ cdecl -arch=win64 ?_Exp@?$_Ctraits@N@std@@SAFPEANNF@Z(ptr double long) _Exp
+@ cdecl -arch=win32 ?_Exp@?$_Ctraits@O@std@@SAFPAOOF@Z(ptr double long) _Exp
+@ cdecl -arch=win64 ?_Exp@?$_Ctraits@O@std@@SAFPEAOOF@Z(ptr double long) _Exp
 @ cdecl -arch=win32 ?_Ffmt@?$num_put@DV?$ostreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@ABAPADPADDH@Z(ptr ptr long long) num_put_char__Ffmt
 @ cdecl -arch=win64 ?_Ffmt@?$num_put@DV?$ostreambuf_iterator@DU?$char_traits@D@std@@@std@@@std@@AEBAPEADPEADDH@Z(ptr ptr long long) num_put_char__Ffmt
 @ cdecl -arch=win32 ?_Ffmt@?$num_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@ABAPADPADDH@Z(ptr ptr long long) num_put_wchar__Ffmt
@@ -3233,9 +3233,9 @@
 @ extern ?_Index@ios_base@std@@0HA ios_base_Index
 @ cdecl -arch=win32 ?_Index_func@ios_base@std@@CAAAHXZ() ios_base_Index_func
 @ cdecl -arch=win64 ?_Index_func@ios_base@std@@CAAEAHXZ() ios_base_Index_func
-@ stub ?_Infv@?$_Ctraits@M@std@@SAMM@Z
-@ stub ?_Infv@?$_Ctraits@N@std@@SANN@Z
-@ stub ?_Infv@?$_Ctraits@O@std@@SAOO@Z
+@ cdecl ?_Infv@?$_Ctraits@M@std@@SAMM@Z(float) std_Ctraits_float__Infv
+@ cdecl ?_Infv@?$_Ctraits@N@std@@SANN@Z(double) std_Ctraits_double__Infv
+@ cdecl ?_Infv@?$_Ctraits@O@std@@SAOO@Z(double) std_Ctraits_double__Infv
 @ stub -arch=arm ?_Init@?$_Mpunct@D@std@@IAAXABV_Locinfo@2@_N@Z
 @ stub -arch=i386 ?_Init@?$_Mpunct@D@std@@IAEXABV_Locinfo@2@_N@Z
 @ stub -arch=win64 ?_Init@?$_Mpunct@D@std@@IEAAXAEBV_Locinfo@2@_N@Z
@@ -3436,9 +3436,9 @@
 @ cdecl -arch=win64 ?_Iput@?$num_put@GV?$ostreambuf_iterator@GU?$char_traits@G@std@@@std@@@std@@AEBA?AV?$ostreambuf_iterator@GU?$char_traits@G@std@@@2@V32@AEAVios_base@2@GPEAD_K@Z(ptr ptr ptr ptr long ptr long) num_put_short__Iput
 @ cdecl -arch=win32 ?_Iput@?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@ABA?AV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@2@V32@AAVios_base@2@_WPADI@Z(ptr ptr long ptr ptr long ptr long) num_put_wchar__Iput
 @ cdecl -arch=win64 ?_Iput@?$num_put@_WV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@std@@@std@@AEBA?AV?$ostreambuf_iterator@_WU?$char_traits@_W@std@@@2@V32@AEAVios_base@2@_WPEAD_K@Z(ptr ptr ptr ptr long ptr long) num_put_wchar__Iput
-@ stub ?_Isinf@?$_Ctraits@M@std@@SA_NM@Z
-@ stub ?_Isinf@?$_Ctraits@N@std@@SA_NN@Z
-@ stub ?_Isinf@?$_Ctraits@O@std@@SA_NO@Z
+@ cdecl ?_Isinf@?$_Ctraits@M@std@@SA_NM@Z(float) std_Ctraits_float__Isinf
+@ cdecl ?_Isinf@?$_Ctraits@N@std@@SA_NN@Z(double) std_Ctraits_double__Isinf
+@ cdecl ?_Isinf@?$_Ctraits@O@std@@SA_NO@Z(double) std_Ctraits_double__Isinf
 @ cdecl ?_Isnan@?$_Ctraits@M@std@@SA_NM@Z(float) std_Ctraits_float__Isnan
 @ cdecl ?_Isnan@?$_Ctraits@N@std@@SA_NN@Z(double) std_Ctraits_double__Isnan
 @ cdecl ?_Isnan@?$_Ctraits@O@std@@SA_NO@Z(double) std_Ctraits_long_double__Isnan
@@ -3535,9 +3535,9 @@
 @ cdecl -arch=arm ?_Myptr@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@IBAPB_WXZ(ptr) basic_string_wchar_const_ptr
 @ thiscall -arch=i386 ?_Myptr@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@IBEPB_WXZ(ptr) basic_string_wchar_const_ptr
 @ cdecl -arch=win64 ?_Myptr@?$basic_string@_WU?$char_traits@_W@std@@V?$allocator@_W@2@@std@@IEBAPEB_WXZ(ptr) basic_string_wchar_const_ptr
-@ stub ?_Nanv@?$_Ctraits@M@std@@SAMM@Z
-@ stub ?_Nanv@?$_Ctraits@N@std@@SANN@Z
-@ stub ?_Nanv@?$_Ctraits@O@std@@SAOO@Z
+@ cdecl ?_Nanv@?$_Ctraits@M@std@@SAMM@Z(float) std_Ctraits_float__Nanv
+@ cdecl ?_Nanv@?$_Ctraits@N@std@@SANN@Z(double) std_Ctraits_double__Nanv
+@ cdecl ?_Nanv@?$_Ctraits@O@std@@SAOO@Z(double) std_Ctraits_double__Nanv
 @ cdecl -arch=arm ?_Narrow_s@?$ctype@D@std@@QBAPBDPBD0DPADI@Z(ptr ptr ptr long ptr long) ctype_char__Narrow_s
 @ thiscall -arch=i386 ?_Narrow_s@?$ctype@D@std@@QBEPBDPBD0DPADI@Z(ptr ptr ptr long ptr long) ctype_char__Narrow_s
 @ cdecl -arch=win64 ?_Narrow_s@?$ctype@D@std@@QEBAPEBDPEBD0DPEAD_K@Z(ptr ptr ptr long ptr long) ctype_char__Narrow_s
@@ -3692,9 +3692,9 @@
 @ cdecl -arch=arm ?_Sgetn_s@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@QAAHPA_WIH@Z(ptr ptr long long) basic_streambuf_wchar__Sgetn_s
 @ thiscall -arch=i386 ?_Sgetn_s@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@QAEHPA_WIH@Z(ptr ptr long long) basic_streambuf_wchar__Sgetn_s
 @ cdecl -arch=win64 ?_Sgetn_s@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@QEAA_JPEA_W_K_J@Z(ptr ptr long long) basic_streambuf_wchar__Sgetn_s
-@ stub ?_Sinh@?$_Ctraits@M@std@@SAMMM@Z
-@ stub ?_Sinh@?$_Ctraits@N@std@@SANNN@Z
-@ stub ?_Sinh@?$_Ctraits@O@std@@SAOOO@Z
+@ cdecl ?_Sinh@?$_Ctraits@M@std@@SAMMM@Z(float float) _FSinh
+@ cdecl ?_Sinh@?$_Ctraits@N@std@@SANNN@Z(double double) _Sinh
+@ cdecl ?_Sinh@?$_Ctraits@O@std@@SAOOO@Z(double double) _Sinh
 @ extern -arch=arm ?_Stinit@?1??_Init@?$basic_filebuf@DU?$char_traits@D@std@@@std@@IAAXPAU_iobuf@@W4_Initfl@23@@Z@4HA basic_filebuf_char__Init__Stinit
 @ extern -arch=i386 ?_Stinit@?1??_Init@?$basic_filebuf@DU?$char_traits@D@std@@@std@@IAEXPAU_iobuf@@W4_Initfl@23@@Z@4HA basic_filebuf_char__Init__Stinit
 @ extern -arch=win64 ?_Stinit@?1??_Init@?$basic_filebuf@DU?$char_traits@D@std@@@std@@IEAAXPEAU_iobuf@@W4_Initfl@23@@Z@4HA basic_filebuf_char__Init__Stinit
@@ -7861,14 +7861,14 @@
 @ cdecl -arch=arm ?xsputn@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@MAAHPB_WH@Z(ptr ptr long) basic_streambuf_wchar_xsputn
 @ thiscall -arch=i386 ?xsputn@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@MAEHPB_WH@Z(ptr ptr long) basic_streambuf_wchar_xsputn
 @ cdecl -arch=win64 ?xsputn@?$basic_streambuf@_WU?$char_traits@_W@std@@@std@@MEAA_JPEB_W_J@Z(ptr ptr long) basic_streambuf_wchar_xsputn
-@ stub _Cosh
+@ cdecl _Cosh(double double)
 @ extern _Denorm
 @ stub _Dnorm
 @ cdecl _Dscale(ptr long)
 @ cdecl _Dtest(ptr)
 @ extern _Eps
 @ cdecl _Exp(ptr double long)
-@ stub _FCosh
+@ cdecl _FCosh(float float)
 @ extern _FDenorm
 @ stub _FDnorm
 @ cdecl _FDscale(ptr long)
@@ -7878,7 +7878,7 @@
 @ extern _FInf
 @ extern _FNan
 # extern _FRteps
-@ stub _FSinh
+@ cdecl _FSinh(float float)
 @ extern _FSnan
 # extern _FXbig
 @ cdecl -norelay _Getcoll()
@@ -7904,15 +7904,15 @@
 # extern _LXbig
 @ extern _LZero
 @ cdecl _Mbrtowc(ptr ptr long ptr ptr)
-@ stub _Mtxdst
-@ stub _Mtxinit
-@ stub _Mtxlock
-@ stub _Mtxunlock
+@ cdecl _Mtxdst(ptr)
+@ cdecl _Mtxinit(ptr)
+@ cdecl _Mtxlock(ptr)
+@ cdecl _Mtxunlock(ptr)
 @ extern _Nan
-@ stub _Once
+@ cdecl _Once(ptr ptr)
 @ stub _Poly
 # extern _Rteps
-@ stub _Sinh
+@ cdecl _Sinh(double double)
 @ extern _Snan
 @ cdecl _Stod(ptr ptr long)
 @ cdecl _Stodx(ptr ptr long ptr)
